aptos-move/framework/aptos-stdlib/sources/cryptography/curves.move

```diff
@@ -0,0 +1,217 @@
+    struct Scalar<phantom Group> has drop {
+        handle: u8
+    }
+
+    struct Point<phantom Group> has drop {
+        handle: u8
+    }
+
+    /// Get internal handle for a Scalar. Currently needed by groth16 module.
+    /// TODO: can this be avoided?
+    public fun get_scalar_handle<Group>(s: &Scalar<Group>): u8 {
+        s.handle
+    }
+
+    /// Get internal handle for a point. Currently needed by groth16 module.
+    /// TODO: can this be avoided?
+    public fun get_point_handle<Group>(p: &Point<Group>): u8 {
+        p.handle
+    }
+
+    /// Perform a bilinear mapping.
+    /// TODO: is it possible to have 2+ mappings between same (G1,G2,Gt)? If so we need a parameter for `mapping_id`?
+    public fun pairing<G1,G2,Gt>(point_1: &Point<G1>, point_2: &Point<G2>): Point<Gt> {
+        Point<Gt> {
+            handle: pairing_internal(point_1.handle, point_2.handle, get_pairing_id<G1,G2,Gt>())
+        }
+    }
+
+    /// Scalar basics.
+    public fun scalar_zero<G>(): Scalar<G> {
+        Scalar<G> {
+            handle: scalar_zero_internal(get_group_id<G>())
+        }
+    }
+
+    public fun scalar_one<G>(): Scalar<G> {
+        Scalar<G> {
+            handle: scalar_one_internal(get_group_id<G>())
+        }
+    }
+
+    public fun scalar_from_u64<G>(value: u64): Scalar<G> {
+        Scalar<G> {
+            handle: scalar_from_u64_internal(value, get_group_id<G>())
+        }
+    }
+
+    public fun scalar_neg<T>(_scalar_1: &Scalar<T>): Scalar<T> {
+        Scalar<T> {
+            handle: scalar_neg_internal(get_group_id<T>())
+        }
+    }
+
+    public fun scalar_add<T>(_scalar_1: &Scalar<T>, _scalar_2: &Scalar<T>): Scalar<T> {
+        Scalar<T> {
+            handle: scalar_add_internal(_scalar_1.handle, _scalar_2.handle, get_group_id<T>())
+        }
+    }
+
+    public fun scalar_mul<T>(_scalar_1: &Scalar<T>, _scalar_2: &Scalar<T>): Scalar<T> {
+        Scalar<T> {
+            handle: scalar_mul_internal(_scalar_1.handle, _scalar_2.handle, get_group_id<T>())
+        }
+    }
+
+    public fun scalar_inv<T>(scalar: &Scalar<T>): Scalar<T> {
+        Scalar<T> {
+            handle: scalar_inv_internal(scalar.handle, get_group_id<T>())
+        }
+    }
+
+    public fun scalar_eq<T>(scalar_1: &Scalar<T>, scalar_2: &Scalar<T>): bool {
+        scalar_eq_internal(scalar_1.handle, scalar_2.handle, get_group_id<T>())
+    }
+
+    public fun bytes_into_scalar<T>(bytes: vector<u8>): Scalar<T> {
+        Scalar<T> {
+            handle: bytes_into_scalar_internal(bytes, get_group_id<T>())
+        }
+    }
+
+    // Point basics.
+    public fun point_identity<T>(): Point<T> {
+        Point<T> {
+            handle: point_identity_internal(get_group_id<T>())
+        }
+    }
+
+    public fun point_generator<T>(): Point<T> {
+        Point<T> {
+            handle: point_generator_internal(get_group_id<T>())
+        }
+    }
+
+    public fun point_add<T>(point_1: &Point<T>, point_2: &Point<T>): Point<T> {
```

Comment:
And we'll want `*_assign` functions too which add/multiply/double/etc in place (see `ristretto255.move`)

Reply:
Unnecessary as developers can write `let x = element_add(x, y)`?

Reply:
Performance wise, my local experiments show there's no difference between `x` and `x_assign`.

Reply:
How did you measure performance in Move? :-/

Reply:
sorry i was talking about `x()` and `x_assign()` in rust.

Reply:
I see. We'll just keep `x()` then, since that makes all our `Element`'s immutable, simplifies the library, and we can easily do `a = x(&a, &a)` in Move to emulate an `x_assign`